applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp

```diff
@@ -32,129 +32,197 @@ namespace Kratos
-    std::vector<double> remote_sum{0, 0, 0, 0, 0};
-    r_comm.SumAll(local_data, remote_sum);
-
-    pos_vol = remote_sum[0];
-    neg_vol = remote_sum[1];
-    inter_area = remote_sum[2];
-    net_inflow_inlet = remote_sum[3];
-    net_inflow_outlet = remote_sum[4];
+    mInterfaceArea = inter_area;
 
     // making a "time step forwards" and updating the
     const double current_time = mrModelPart.GetProcessInfo()[TIME];
     const double current_dt = mrModelPart.GetProcessInfo()[DELTA_TIME];
-    mQNet2 = mQNet1;
-    mQNet1 = mQNet0;
-    mQNet0 = net_inflow_inlet + net_inflow_outlet;
 
-    // adding time-integrated net inflow ( = new water volume ) to the theoretical value
-    // The value of the integral is evaluated following the idea of the Adams-Moulton-Procedure (s=2)
-    mTheoreticalNegativeVolume += current_dt * ( 5.0*mQNet0 + 8.0*mQNet1 - 1.0*mQNet2 ) / 12.0;
+    mQNet0 = net_inflow_inlet + net_inflow_outlet;
+    mTheoreticalNegativeVolume += current_dt * mQNet0;
 
-    // calculation of the "missing" water volume
-    const double water_volume_error = mTheoreticalNegativeVolume - neg_vol;
 
-    double shift_for_correction = 0.0;
-    // check if it is time for a correction (if wished for)
-    if ( mPerformCorrections && mrModelPart.GetProcessInfo()[STEP] % mCorrectionFreq == 0 && inter_area > 10e-7){
-        // if water is missing, a shift into negative direction increases the water volume
-        shift_for_correction = - water_volume_error / inter_area;
-        ShiftDistanceField( shift_for_correction );
-    }
+    mWaterVolumeError = mTheoreticalNegativeVolume - neg_vol;
 
     // assembly of the log message
     std::string output_line_timestep =  std::to_string(current_time) + "\t\t";
     output_line_timestep +=             std::to_string(neg_vol) + "\t\t";
     output_line_timestep +=             std::to_string(pos_vol) + "\t\t";
-    output_line_timestep +=             std::to_string(water_volume_error) + "\t\t";
+    output_line_timestep +=             std::to_string(mWaterVolumeError) + "\t\t";
     output_line_timestep +=             std::to_string(net_inflow_inlet) + "\t\t";
     output_line_timestep +=             std::to_string(net_inflow_outlet) + "\t\t";
-    output_line_timestep +=             std::to_string(inter_area) + "\t\t";
-    output_line_timestep +=             std::to_string(shift_for_correction) +"\n";
+    output_line_timestep +=             std::to_string(inter_area) + "\n";
+
     return output_line_timestep;
 }
 
 
 
-void MassConservationCheckProcess::ComputeVolumesAndInterface( double& positiveVolume, double& negativeVolume, double& interfaceArea ){
+double MassConservationCheckProcess::ComputeDtForConvection(){
+
+    // a small step is set to avoid numerical problems
+    double time_step_for_convection = 1.0e-7;
+    const auto& r_comm = mrModelPart.GetCommunicator();
+
+    if ( mWaterVolumeError > 0.0 ){
```

Comment:
@philbucher This is synched at the moment it is computed. Should I sync it again here?

Reply:
Is it needed to be a process? No single method of a process is used, it looks like a utility to me. This is something you cannot add to your list of processes and expect to do anything.

Reply:
I completely agree.

Reply:
Right now, the process-like API is in the `apply_mass_conservation_check_process.py` (this is why the c++ implementation doesn't follow it).

As a comment aside, I personally think that these kind of utils shouldn't be exposed to the user but encapsulated inside the solver (as we do with the level set correction in the embedded solvers). This allows to set the most suitable settings in the `solver_settings` defaults for non-experienced users but keeping the possibility of customizing the settings for advanced ones. Besides this, it simplifies the input json settings as this process is always required.

Having said this, we can refactor it to be a utility, as @jrubiogonzalez points, or to have the `Process` API also in the c++ implementation. The latter would be more flexible in the sense that it would make possible to eventually use it in combination with another solver by adding it to the process list (OFC once we finally implement the Processes register).

Reply:
I think the best option is to convert the cpp in a real independent process from python side.

Reply:
Mm actually, I think its better to do it from python as we need to call the levelset. Then I guess cpp file fits better as an utility.